v-next/core/test/plugins/resolve-plugin-list.ts

```diff
@@ -0,0 +1,119 @@
+
+describe("Plugins - resolve plugin list", () => {
+  it("should return empty on an empty plugin list", () => {
+    assert.deepStrictEqual(resolvePluginList([]), []);
+  });
+
+  it("should return empty on an undefined plugin list", () => {
+    assert.deepStrictEqual(resolvePluginList(), []);
+  });
+
+  it("should return a single plugin", () => {
+    const plugin: HardhatPlugin = {
+      id: "example-plugin",
+    };
+
+    assert.deepStrictEqual(resolvePluginList([plugin]), [plugin]);
+  });
+
+  it("should support nested dependencies", () => {
+    // A -> B -> C
+    const c = { id: "c" };
+    const b = { id: "b", dependencies: [c] };
+    const a = { id: "a", dependencies: [b] };
+
+    assert.deepStrictEqual(resolvePluginList([a]), [c, b, a]);
+  });
+
+  it("should break ties by honouring array order", () => {
+    // A / B / C
+    const c = { id: "c" };
+    const b = { id: "b" };
+    const a = { id: "a" };
+
+    assert.deepStrictEqual(resolvePluginList([a, b, c]), [a, b, c]);
+  });
+
+  it("should break ties by honouring subdependency array order", () => {
+    //   A
+    //  / \
+    // B   C
+    const c = { id: "c" };
+    const b = { id: "b" };
+    const a = { id: "a", dependencies: [b, c] };
+
+    assert.deepStrictEqual(resolvePluginList([a]), [b, c, a]);
+  });
+
+  it("should support shared dependencies", () => {
+    // A   B
+    //  \ /
+    //   C
+    const c = { id: "c" };
+    const b = { id: "b", dependencies: [c] };
+    const a = { id: "a", dependencies: [c] };
+
+    assert.deepStrictEqual(resolvePluginList([a, b]), [c, a, b]);
+  });
+
+  it("should support shared subdependencies", () => {
+    //   A
+    //  / \
+    // B   C
+    //  \ /
+    //   D
+    const d = { id: "d" };
+    const c = { id: "c", dependencies: [d] };
+    const b = { id: "b", dependencies: [d] };
+    const a = { id: "a", dependencies: [b, c] };
+
+    assert.deepStrictEqual(resolvePluginList([a]), [d, b, c, a]);
+  });
+
+  it("should deal with a complicated dependency graph", () => {
+    //   A    B
+    //  / \  / \ \
+    // C   D    E F
+    //  \ /     \/
+    //   G      H
+    //    \    /
+    //      I
+    const i = { id: "i" };
+    const h = { id: "h", dependencies: [i] };
+    const g = { id: "g", dependencies: [i] };
+    const f = { id: "f", dependencies: [h] };
+    const e = { id: "e", dependencies: [h] };
+    const d = { id: "d", dependencies: [g] };
+    const c = { id: "c", dependencies: [g] };
+    const b = { id: "b", dependencies: [d, e, f] };
+    const a = { id: "a", dependencies: [c, d] };
+
+    assert.deepStrictEqual(resolvePluginList([a, b]), [
+      i,
+      g,
+      c,
+      d,
+      a,
+      h,
+      e,
+      f,
+      b,
+    ]);
+  });
+
+  it("should error on different plugins with the same id", () => {
+    const a = { id: "dup" };
+    const copy = { id: "dup" };
+
+    assert.throws(
+      () => resolvePluginList([a, copy]),
```

Comment:
It should throw a `HardhatError` and test that instead of the string.

Reply:
I added the Hardhat error for duplicate plugins under `GENERAL` in the error descriptors. Let me know if that is not the category you were expecting.